geomstats/information_geometry/normal.py

```diff
@@ -116,8 +160,420 @@ def pdf(x):
+            of covariance matrices.
+        """
+        mean, diagonal = self._unstack_mean_diagonal(self.sample_dim, point)
+        regularized = gs.where(diagonal < gs.atol, gs.atol, diagonal)
+        projected = self._stack_mean_diagonal(mean, regularized)
+        return projected
+
+    def sample(self, point, n_samples=1):
+        """Sample from the diagonal multivariate normal distribution.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 2 * sample_dim]
+            Point on the manifold. First :math:`sample_dim` elements contain
+            the mean vector and the last :math:`sample_dim` elements contain
+            the diagonal of the covariance matrix.
+        n_samples : int
+            Number of points to sample with each pair of parameters in point.
+            Optional, default: 1.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., n_samples, sample_dim]
+            Sample from multivariate normal distributions.
+        """
+        geomstats.errors.check_belongs(point, self)
+        if point.ndim > 2:
+            raise NotImplementedError
+        point = gs.to_ndarray(point, to_ndim=2)
+        samples = []
+        for p in point:
+            mean, diag = self._unstack_mean_diagonal(self.sample_dim, p)
+            cov = gs.vec_to_diag(diag)
+            samples.append(gs.array(multivariate_normal.rvs(mean, cov, size=n_samples)))
+        return samples[0] if point.shape[0] == 1 else gs.stack(samples)
+
+    def point_to_pdf(self, point):
+        """Compute pdf associated to point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 2 * sample_dim]
+            Point representing a probability distribution.
+            First :math:`sample_dim` elements contain the mean vector and the last
+            :math:`sample_dim` elements contain the diagonal of the covariance matrix.
+
+        Returns
+        -------
+        pdf : function
+            Probability density function of the normal distribution with
+            parameters provided by point.
+        """
+        geomstats.errors.check_belongs(point, self)
+        if point.ndim > 2:
+            raise NotImplementedError
+        point = gs.to_ndarray(point, to_ndim=2, axis=0)
+        point = point[:, None, :]
+        n = self.sample_dim
+        mean, diagonal = self._unstack_mean_diagonal(n, point)
+
+        def pdf(x):
+            """Generate parameterized function for normal pdf.
+
+            Parameters
+            ----------
+            x : array-like, shape=[n_samples, sample_dim]
+                Points at which to compute the probability
+                density function.
+            """
+            x = gs.to_ndarray(x, to_ndim=2, axis=0)
+            x = x[None, :, :]
+            det_cov = gs.squeeze(gs.prod(diagonal, axis=-1))
+            pdf_normalization = 1 / gs.sqrt(gs.power((2 * gs.pi), n) * det_cov)
```

Comment:
independent of `x`. can be moved outside.

Reply:
This generates error 'pdf_normalization referenced before assignment' when moved outside.